test/subdomainregistrar/TestRentalSubdomainRegistrar.js

```diff
@@ -0,0 +1,778 @@
+      await BaseRegistrar.setApprovalForAll(NameWrapper.address, true)
+      await NameWrapper.wrapETH2LD(
+        'test',
+        account,
+        CANNOT_UNWRAP,
+        EMPTY_ADDRESS,
+      )
+      expect(await NameWrapper.ownerOf(node)).to.equal(account)
+    })
+
+    it('should emit an event when setting up a domain', async () => {
+      const tx = await SubdomainRegistrar.setupDomain(
+        node,
+        RentalPricer.address,
+        account,
+        true,
+      )
+
+      await expect(tx)
+        .to.emit(SubdomainRegistrar, 'NameSetup')
+        .withArgs(node, RentalPricer.address, account, true)
+    })
+  })
+
+  describe('register', () => {
+    beforeEach(async () => {
+      await BaseRegistrar.register(labelhash('test'), account, 86400 * 2)
+      await BaseRegistrar.setApprovalForAll(NameWrapper.address, true)
+      await NameWrapper.wrapETH2LD(
+        'test',
+        account,
+        CANNOT_UNWRAP,
+        EMPTY_ADDRESS,
+      )
+      expect(await NameWrapper.ownerOf(node)).to.equal(account)
+      await SubdomainRegistrar.setupDomain(
+        node,
+        RentalPricer.address,
+        account,
+        true,
+      )
+    })
+    it('should allow subdomains to be created', async () => {
+      const balanceBefore = await Erc20WithAccount2.balanceOf(account2)
+      const duration = 86400
+      const [, fee] = await RentalPricer.price(
+        namehash('test.eth'),
+        'subname',
+        duration,
+      )
+      await NameWrapper.setApprovalForAll(SubdomainRegistrar.address, true)
+
+      await Erc20WithAccount2.approve(
+        SubdomainRegistrar.address,
+        ethers.constants.MaxUint256,
+      )
+
+      await SubdomainRegistrar2.register(
+        node,
+        'subname',
+        account2,
+        EMPTY_ADDRESS,
+        0,
+        duration,
+        [],
+      )
+      const block = await ethers.provider.getBlock('latest')
+      const balanceAfter = await Erc20WithAccount2.balanceOf(account2)
+      expect(balanceBefore.sub(balanceAfter)).to.equal(fee)
+      const [owner, fuses, expiry] = await NameWrapper.getData(subNode)
+      expect(owner).to.equal(account2)
+      expect(expiry).to.equal(block.timestamp + duration)
+      expect(fuses).to.equal(PARENT_CANNOT_CONTROL)
+    })
+    it('should not allow subdomains to be created on unapproved parents', async () => {
+      await Erc20.approve(
+        SubdomainRegistrar.address,
+        ethers.constants.MaxUint256,
+      )
+      await expect(
+        SubdomainRegistrar.register(
+          node,
+          'subname',
+          account2,
+          EMPTY_ADDRESS,
+          0,
+          86400,
+          [],
+        ),
+      ).to.be.revertedWith(
+        `Unauthorised("${namehash('test.eth')}", "${
+          SubdomainRegistrar.address
+        }")`,
+      )
+    })
+
+    it('should allow subdomains to be registered without a fee', async () => {
```

Comment:
Should't the same test exist on `TestForeverSubdomainRegistrar.js`?

Reply:
Or extract common functions between rental and forever registrar into seperate test, or use `shouldBehaveLike*` pattern you used for NameWrapper tests.